<p style="background:#eeeeee">
Informática - 1º de Física
<br>
<strong>Introducción a la Programación</strong>
</p>

## Tema 1. Primeros pasos

El punto de partida es instalar en nuestro ordenador el lenguaje de programación que vamos a utilizar y algunas herramientas auxiliares. Una vez hecho esto, existen diferentes formas de trabajar:

- Editor y terminal.

- IDE (entorno integrado de programación).

- Notebook interactivo.

En su momento haremos ejercicios para familiarizarnos con las dos primeras pero la mayoría de las explicaciones las haremos utilizando los *jupyter notebooks*. De esta manera podemos crear documentos de referencia con todo lo que vamos aprendiendo.

A partir de aquí suponemos que conocemos los comandos de edición básicos de los notebooks, que se explican en la primera clase de prácticas.

Se recomienda crear una carpeta en tu ordenador para guardar todos los documentos que se vayan creando en la asignatura.

### Aritmética elemental

Igual que una calculadora, un ordenador es capaz de hacer operaciones aritméticas:

In [1]:
3+2

5

Los símbolos de las operaciones son más o menos intuitivos: `+`, `-`, `*`, `/`, y la precedencia es la usual en matemáticas.

In [2]:
2 + 3 * 5

17

Los espacios no influyen. Usamos paréntesis para agrupar de la forma deseada:

In [3]:
(2+3) *  5-4

21

La potencia se expresa con el símbolo `**`.

In [4]:
2**10

1024

En programación es muy importante distinguir entre números enteros y números reales. Los enteros son matemáticamente exactos, mientras que los reales son aproximaciones con una cantidad finita de decimales.

In [5]:
2**(1/2)

1.4142135623730951

En algunos lenguajes los enteros tienen un tamaño limitado, como ocurre en las calculadoras. Pero en Python los enteros pueden ser de cualquier tamaño. El único límite es la memoria disponible en el ordenador.

In [6]:
13**200

6147102592468651336192412089282069174074558674790952240521262267958019995940288912789820651282318338938794713404346722691907328564818184457431599939243408750167870757463938171866175799570066146915175087199725900885179188001

El símbolo de división `/` se refiere a la división real:

In [7]:
10 / 2

5.0

In [8]:
11 / 2

5.5

El cociente y el resto de la división entera utilizan respectívamente los símbolos `//` y `%`: 

In [9]:
11 // 2

5

In [10]:
11 % 2

1

In [11]:
abs( 5.0 - 8 )

3.0

Los números reales que se utilizan normalmente son aproximaciones de "coma flotante" de "doble precisión" (64bits) que tienen 15 decimales y un exponente hasta $\pm 308$.

La notación científica $a\times 10^b$ se indica con la letra `e`.

In [12]:
2.5e3

2500.0

In [13]:
1e-2

0.01

In [14]:
2.37e18 ** 5.1

5.143000477600881e+93

Debido a que los números reales tienen un número fijo de decimales las operaciones suelen producir errores de redondeo:

In [15]:
10/3

3.3333333333333335

In [16]:
0.1 + 0.2

0.30000000000000004

### Bibliotecas de funciones

El lenguaje Python posee un número enorme de funciones y operaciones disponibles que están organizadas en una gran biblioteca estructurada por "módulos". Por ejemplo, las funciones matemáticas (trigonométricas, logaritmos, etc.) están en el módulo `numpy`.

Para utilizar funciones de un módulo debemos "importarlas" (hacerlas visibles) con una instrucción como la siguiente:

In [17]:
import numpy as np

In [18]:
np.sin(np.pi/6)

0.49999999999999994

In [19]:
np.cos(0)

1.0

In [20]:
np.log(10)

2.302585092994046

Es muy conveniente estar familiarizado con las funciones disponibles en los módulos relacionados con nuestro ámbito de aplicación (principalmente `numpy`, `scipy`, `matplotlib` y `pandas`). No hay que aprender nada de memoria sino tener a mano la documentación, que está disponible online  (p. ej. [aquí tenemos](
https://docs.python.org/3/library/math.html) las funciones disponibles en `math`) y localmente con la función `help`. En el menú "Help" del notebook hay enlaces a los módulos más importantes. En cualquier caso una búsqueda en google nos lleva rápidamente a la función deseada.

Con lo visto hasta ahora podemos usar el lenguaje Python como una calculadora. A partir de aquí iremos introduciendo la capacidad de programar las operaciones. 

### Secuencia de instrucciones

Python evalúa una detrás de otra todas las expresiones que encuentra. Pero el notebook solo muestra el resultado de la última que aparece en la celda:

In [21]:
1 + 1

2 + 3

5 * 2

10

Si queremos ver todos los resultados utilizamos la función `print`:

In [22]:
print(1 + 1)

print(2 + 3)

print(5 * 2)

2
5
10


### Nombres

Un aspecto fundamental en la programación consiste en ponerle nombres a los números y otros tipos de datos con los que estamos trabajando. Para ello, Python (y la mayoría de lenguajes) usa el símbolo `=`. Por ejemplo, si hacemos lo siguiente

In [23]:
a = 5

z = 1/2 + 3*a

entonces los nombres `a` y `z` queda definidos con los siguientes valores:

In [24]:
a

5

In [25]:
z

15.5

La elección del signo `=` no es muy afortunada, ya que esta acción no tiene mucho que ver con el concepto de igualdad matemática. (Otros lenguajes usan signos más intuitivos como `<-` o `:=` para esta operación.)

Cuando veamos expresiones como `z = 1/2 + 3*a`  no tenemos que pensar en ningún momento en ecuaciones matemáticas. Nada más lejos de la realidad. Simplemente se da un nombre al **resultado** de evaluar la expresión. Podríamos decir que estamos "definiendo" un nombre, pero este término también puede inducir a confusión, ya que el nombre queda asignado al valor final de la expresión, y no a la expresión en sí. Cada vez que aparezca `z` en el programa se sustituirá automáticamente por `15.5`. No se sustituye por `1/2 + 3*a`. 

Este matiz es importante debido a que los nombres se pueden redefinir. Observa la siguiente secuencia de definiciones:

In [26]:
a = 5

b = 2*a

a = 7

b

10

Recuerda que las expresiones se evalúan una detrás de otra. La última de ellas, `b` tiene el valor `10` a pesar de que se ha definido como `b = 2*a` y que justo antes se ha redefinido `a` como `a=7`. El cambio en un nombre no afecta a los demás que dependen de él. No se recalcula nada. 

Más adelante explicaremos cómo hacer verdaderas definiciones en las que unos símbolos dependen de otros. Por el momento podemos pensar en la idea de almacenar en memoria un dato que podemos recuperar cómodamente mediante un nombre.

En algunos lenguajes se dice que hemos "almacenado un valor en una variable", o que hemos "asignado un valor". Tampoco es lo ideal porque la palabra "variable" puede inducir a confusión con las variables matemáticas. (Y no causa buena impresión almacenar constantes como por ejemplo el número $\pi$ dentro variables...)

No hay ningún problema en redefinir una variable usando su valor anterior. Esto puede dar lugar a expresiones aparentemente absurdas como:

In [27]:
b = b + 5

Simplemente tenemos que recordar que lo anterior no es una ecuación sino una **redefinición**, cuyo efecto es sustituir del valor de `b`, que antes era `10`, por `15`. 

Esta operación se puede abreviar como `b += 5`.

Cuando se redefine un nombre el valor que tenía previamente se olvida de forma irrecuperable.

Para hacer una definición el nombre (cualquier palabra que empiece por una letra) se pone a la izquierda del signo `=` y a la derecha se pone la expresión deseada, donde pueden aparecer nombres previamente definidos. Por tanto, lo siguiente es incorrecto:

In [28]:
1 + 1 = c

<class 'SyntaxError'>: cannot assign to expression here. Maybe you meant '==' instead of '='? (<ipython-input-28-bfcb07cbd8e2>, line 1)

In [29]:
c = 3*x + 5

<class 'NameError'>: name 'x' is not defined

Es conveniente elegir nombres descriptivos o abreviaturas para indicar el papel que juega cada dato en el programa.

In [30]:
masa = 10
velocidad_inicial = 8.5
pos = 5

α  = np.pi/4

Es una cuestión de gusto personal, pero siempre se recomienda la claridad y la sencillez.

### Estado de la computación

Cuando el ordenador "ejecuta" un programa (o Python evalúa celdas de código) ocurre lo siguiente:

- En todo momento existe un *estado* de la computación, que es simplemente un conjunto de datos con sus correspondientes nombres (inicialmente vacío).

- Algunas instrucciones cambian elementos del estado. Otras no, aunque sí actúan sobre el "mundo exterior" (por ejemplo, mostrando algo en la pantalla). El efecto producido por una instrucción depende del estado anterior.

Una computación es la secuencia de estados producida por las sucesivas instrucciones del programa. El *resultado* de la computación es el estado final.

Por ejemplo:

In [31]:
# Instrucción       Estado
                #   {}
a = 5
                #   { a:5 }
b = 2*a
                #   { a:5 , b:10 }
print(b)
                #   { a:5 , b:10 }
a = 7
                #   { a:7 , b:10 }
b += 2
                #   { a:7 , b:12 }
a+b
                #   { a:7 , b:12 }
a-b
                #   { a:7 , b:12 }

10


-5

La instrucción `print` de la línea 7 escribe `10` en la pantalla (o notebook). La línea 13 realiza una operación pero no tiene ningún efecto observable. El resultado se descarta. La línea 15 sí imprime el resultado `-5` por ser la última (esto es un convenio de los notebooks para ahorrar la instrucción `print`).

Los herramientas de programación suelen tener una ventana para mostrar en todo momento el valor de las variables definidas. En los jupyter notebooks esto se consigue con la extensión "variable inspector".

### Tipos de datos

Un concepto esencial en programación es el de **tipo** de un dato. Intuitivamente, podemos pensar en un conjunto de valores posibles. Un buen ejemplo son los tipos numéricos, que tratan de imitar a los conjuntos matemáticos usuales $\mathbb Z$, $\mathbb Q$, $\mathbb R$, $\mathbb C$, etc.

Hasta el momento hemos visto datos enteros (tipo `int`) y reales (tipo `float`) pero los ordenadores pueden 
trabajar con muchos otros tipos de información. La función `type` sirve para consultar de qué tipo es un dato:

In [32]:
type(17)

int

In [33]:
b = -2/3

type(b)

float

In [34]:
type(2+3j)

complex

In [35]:
type('hola')

str

### Redondeo

In [36]:
round(3.6)

4

In [37]:
round( 2**0.5, 3 )

1.414

In [38]:
np.ceil(3.2)

4.0

In [39]:
np.floor(3.9)

3.0

In [40]:
np.sign(-2.7)

-1.0

### Grados y radianes

In [41]:
np.degrees(np.pi/2)

90.0

In [42]:
np.sin(np.radians(30))

0.49999999999999994

### Números complejos

Las constantes imaginarias se introducen añadiendo la letra `j`:

In [43]:
3j ** 2

(-9+0j)

In [44]:
a = 2+3j

np.cos(a)

(-4.189625690968807-9.109227893755337j)

In [45]:
np.cos(a)**2 + np.sin(a)**2

(1-1.4210854715202004e-14j)

In [46]:
np.sqrt(-25)

nan

In [47]:
np.sqrt(-25+0j)

5j

In [48]:
np.sqrt(complex(-25))

5j

### Secuencias (*arrays*) de números

Muchas operaciones matemáticas se pueden efectuar automáticamente con secuencias de números.

In [49]:
np.array([3, 5, -2, 8])

array([ 3,  5, -2,  8])

In [50]:
np.linspace(1, 5, 9)

array([1. , 1.5, 2. , 2.5, 3. , 3.5, 4. , 4.5, 5. ])

In [51]:
np.arange(7)

array([0, 1, 2, 3, 4, 5, 6])

In [52]:
np.arange(2, 8, 0.4)

array([2. , 2.4, 2.8, 3.2, 3.6, 4. , 4.4, 4.8, 5.2, 5.6, 6. , 6.4, 6.8,
       7.2, 7.6])

In [53]:
np.random.rand(20)

array([0.83171625, 0.73327653, 0.13234097, 0.44831189, 0.0663959 ,
       0.13034724, 0.7107223 , 0.60878669, 0.4843882 , 0.87403732,
       0.75462843, 0.54220399, 0.51453889, 0.1868563 , 0.01444807,
       0.54618392, 0.30454023, 0.19871984, 0.92394029, 0.44429383])

In [54]:
k = np.arange(10)
k**2

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])

In [55]:
np.sum(k)

45

### Operaciones lógicas

 `>`, `<`, `>=`, `<=`, `==`, `!=`, `and`, `or`, `not`.

Un tipo de datos esencial en programación es el tipo `bool` (lógico), que solo puede tomar dos valores: `True` o `False`. Aparece como resultado de operaciones de comparación. Por ejemplo:

In [56]:
3 < 10

True

In [57]:
type(1 > 2)

bool

In [58]:
b - 3 == 8 + 7

False

Las operaciones de comparación se expresan con símbolos más o menos autoexplicativos: `>`, `<`, `>=`, `<=`, `==`, `!=` y, como veremos, se usan para tomar decisiones en los programas. Dados dos números comprueban si la condición es cierta o no.

Observa que el símbolo `==` que se usa para comprobar si dos números son iguales se escribe con dos signos de igual seguidos para distinguirlo del operador `=` de asignación.

A pesar de su aspecto, las operaciones de comparación no establecen una relación matemática entre las variables. La sentencia anterior no define una ecuación que el ordenador tenga que resolver. Es simplemente una expresión que se evalúa y produce un resultado que en este caso no es numérico sino lógico (verdadero o falso).

Los valores lógicos se pueden combinar mediante las conectivas `and`, `or`, `not` de la lógica proposicional. Por ejemplo:

In [59]:
3 < 4 or 2+2 == 5

True

In [60]:
b == 13 and (b < 0 or not 2+2==5)

False

La precedencia de los operadores lógicos es la usual: primero se evalúan las comparaciones, a continuación `not`, después `and` y finalmente `or`.

In [61]:
not True and False or True

True

In [62]:
not (True and (False or True))

False

Finalmente, es importante señalar que nunca se deben hacer comparaciones de igualdad exacta con números de tipo `float` aunque lo permita el lenguaje:

In [63]:
0.2 + 0.5 == 0.7

True

In [64]:
0.2 + 0.1 == 0.3

False

Este último resultado inesperado (y erróneo) se debe al carácter aproximado de los números reales informáticos y a la imposibilidad de representar exactamente muchos números racionales en la base de numeración binaria utilizada internamente.
Como veremos después, la comparación de números reales debe hacerse estableciendo una tolerancia sobre su diferencia absoluta o relativa, según convenga.

El procesador opera con los números de coma flotante de una forma muy razonable, con un compromiso entre precisión y velocidad adecuado para la mayoría de las aplicaciones. Pero olvidar la naturaleza aproximada de la computación numérica puede conducir a errores en el programa que se manifiestan cuando menos te lo esperas.




### Cadenas de caracteres

Otro tipo de datos importante es la cadena de caracteres (`str`). Se trata simplemente de una secuencia de letras, números o cualquier otro símbolo. Es un texto entre comillas que en principio no tiene ningún significado para el ordenador. En las aplicaciones científicas se usa normalmente para mostrar mensajes al usuario humano, o para introducir información a los programas. 

In [65]:
nombre = "Alberto"

saludo = 'Hola ' + nombre

saludo

'Hola Alberto'

(Por comodidad algunas operaciones distintas como la suma de números y la concatenación de cadenas se representan con el mismo símbolo.)

Las *F-strings* permiten insertar expresiones y mostrarlas con el formato deseado:

In [66]:
print(F'Hola {nombre}, 2+3={2+3} y el número pi es aproximadamente π = {np.pi:.4f}')

Hola Alberto, 2+3=5 y el número pi es aproximadamente π = 3.1416


Más adelante veremos algunas operaciones útiles para manipular cadenas pero en este curso nos centraremos principalmente en operaciones numéricas.

Finalmente, vamos a presentar brevemente algunos tipos numéricos especializados para hacernos una idea de las herramientas disponibles para la computación científica.

Los apartados marcados con asterisco pueden omitirse en una primera lectura.

### Racionales *

El módulo [fractions](https://docs.python.org/3.6/library/fractions.html) proporciona números racionales exactos. Son útiles en teoría de números pero no se utilizan mucho en ciencias experimentales.

In [67]:
from fractions import Fraction

In [68]:
x = Fraction(13,1001)

In [69]:
x

Fraction(1, 77)

In [70]:
Fraction(7,8)**17 + x

Fraction(20164349390700187, 173388585653764096)

### Precisión arbitraria *

La aritmética de coma flotante con precisión arbitraria se puede conseguir (entre otros) con el módulo [mpmath](http://mpmath.org/). Esto permite elegir el número de decimales deseado en los cálculos.

In [71]:
from mpmath import mp
mp.dps = 100

(1+mp.sqrt(5))/2

mpf('1.618033988749894848204586834365638117720309179805762862135448622705260462818902449707207204189391137492')

La función `mp.mpf` construye números de este tipo para usarlos en la artimética normal. Pero hay que tener cuidado. Observa la diferencia:

In [72]:
mp.mpf(1/3)

mpf('0.333333333333333314829616256247390992939472198486328125')

In [73]:
mp.mpf(1)/3

mpf('0.3333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333345')

En ciencias experimentales no necesitamos normalmente una precisión tan enorme. Además, estos números no llevan la cuenta del número de dígitos exactos que van quedando después de cada operación. 

### Magnitudes físicas *

El módulo [uncertainties](http://pythonhosted.org/uncertainties/) automatiza la propagación de errores. Las magnitudes físicas se pueden manejar mediante [pint](http://pint.readthedocs.io/en/0.8/tutorial.html). Estas herramientas son muy útiles para analizar datos de los laboratorios de física. Más adelante veremos algunos ejemplos de uso.

### Ejercicios

Ver la [Tarea 2](Tarea2.ipynb).